# 04. Embeddings e indexacion en PostgreSQL + pgvector

Este notebook implementa la carga del corpus RAG en PostgreSQL, la generacion de embeddings y la validacion inicial de busqueda semantica y keyword usando la capa productiva de `src.vector_store`.

> Si cambias `EMBEDDING_MODEL`, reejecuta el notebook completo para regenerar los embeddings persistidos en PostgreSQL.

## Objetivos

- cargar el dataset `rag.parquet`
- crear o validar el esquema `pgvector`
- insertar convocatorias por lotes usando `src.vector_store`
- comprobar el numero de registros persistidos
- ejecutar busquedas de ejemplo


In [1]:
from __future__ import annotations

import sys
from pathlib import Path

ROOT = Path.cwd().resolve().parent if Path.cwd().name == "notebooks" else Path.cwd().resolve()
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

import pandas as pd
from IPython.display import display
from sqlalchemy import text

from src.config import PRIMARY_RAG_PATH, RAG_TEXT_COLUMN, settings
from src.data_loader import load_retrieval_corpus
from src.vector_store import (
    connect_db,
    create_schema_if_needed,
    insert_convocatorias,
    keyword_search,
    semantic_search,
)

BATCH_SIZE = 64
SAMPLE_LIMIT: int | None = None

print("ROOT:", ROOT)
print("DATABASE_URL:", settings.database_url)
print("EMBEDDING_MODEL:", settings.embedding_model)
print("RAG_PATH:", PRIMARY_RAG_PATH)


/var/www/codigo/maestria_ia/umsa/diplomados_intermedios/dip_03/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


ROOT: /var/www/codigo/maestria_ia/umsa/diplomados_intermedios/dip_03
DATABASE_URL: postgresql+psycopg://postgres:postgres@localhost:5432/sicoes_rag
EMBEDDING_MODEL: sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2
RAG_PATH: /var/www/codigo/maestria_ia/umsa/diplomados_intermedios/dip_03/data/rag/sicoes_convocatorias_rag.parquet


## 1. Cargar corpus RAG

El corpus indexado parte del dataset RAG canónico versionado en `parquet`.


In [2]:
df = load_retrieval_corpus()
if SAMPLE_LIMIT is not None:
    df = df.head(SAMPLE_LIMIT).copy()

print("Filas cargadas:", len(df))
print("Columnas:", df.columns.tolist())
df[["document_id", "cuce", "entidad", RAG_TEXT_COLUMN]].head(3)


Filas cargadas: 1429
Columnas: ['document_id', 'cuce', 'entidad', 'tipo_contratacion', 'modalidad', 'objeto_contratacion', 'subasta', 'fecha_publicacion', 'fecha_presentacion', 'fecha_publicacion_iso', 'fecha_presentacion_iso', 'estado', 'archivos_disponibles', 'ficha_url', 'source_draw', 'source_row_in_page', 'longitud_objeto', 'cantidad_palabras_objeto', 'texto_rag']


,document_id,cuce,entidad,texto_rag
0,26-1101-00-1669685-1-1,26-1101-00-1669685-1-1,Gobierno Autonomo Municipal De Sucre,CUCE: 26-1101-00-1669685-1-1\nEntidad: Gobiern...
1,26-0417-03-1669697-1-1,26-0417-03-1669697-1-1,Caja Nacional De Salud Regional Santa Cruz,CUCE: 26-0417-03-1669697-1-1\nEntidad: Caja Na...
2,26-0417-03-1669695-1-1,26-0417-03-1669695-1-1,Caja Nacional De Salud Regional Santa Cruz,CUCE: 26-0417-03-1669695-1-1\nEntidad: Caja Na...


## 2. Crear esquema y validar conexion

Esta celda usa el SQL de inicializacion y la configuracion de `src.vector_store`. Si falla en tu entorno, revisa `DATABASE_URL` y el estado del contenedor Docker.


In [3]:
create_schema_if_needed()
engine = connect_db()

with engine.connect() as connection:
    print("SELECT 1 ->", connection.exec_driver_sql("SELECT 1").scalar())
    table_exists = connection.exec_driver_sql(
        "SELECT to_regclass('public.convocatorias')"
    ).scalar()
    print("Tabla convocatorias:", table_exists)


SELECT 1 -> 1
Tabla convocatorias: convocatorias


## 3. Insercion por lotes

La insercion se hace mediante `insert_convocatorias()` con `upsert`, por lo que el notebook puede reejecutarse sin duplicar registros por `document_id`.


In [4]:
records = df.to_dict(orient="records")
total_batches = (len(records) + BATCH_SIZE - 1) // BATCH_SIZE
inserted_total = 0

for batch_index in range(total_batches):
    start = batch_index * BATCH_SIZE
    end = start + BATCH_SIZE
    batch = records[start:end]
    inserted = insert_convocatorias(batch, generate_missing_embeddings=True, upsert=True)
    inserted_total += inserted
    print(f"Lote {batch_index + 1}/{total_batches}: {inserted} registros procesados")

print("Total de registros procesados:", inserted_total)


Lote 1/23: 64 registros procesados
Lote 2/23: 64 registros procesados
Lote 3/23: 64 registros procesados
Lote 4/23: 64 registros procesados
Lote 5/23: 64 registros procesados
Lote 6/23: 64 registros procesados
Lote 7/23: 64 registros procesados
Lote 8/23: 64 registros procesados
Lote 9/23: 64 registros procesados
Lote 10/23: 64 registros procesados
Lote 11/23: 64 registros procesados
Lote 12/23: 64 registros procesados
Lote 13/23: 64 registros procesados
Lote 14/23: 64 registros procesados
Lote 15/23: 64 registros procesados
Lote 16/23: 64 registros procesados
Lote 17/23: 64 registros procesados
Lote 18/23: 64 registros procesados
Lote 19/23: 64 registros procesados
Lote 20/23: 64 registros procesados
Lote 21/23: 64 registros procesados
Lote 22/23: 64 registros procesados
Lote 23/23: 21 registros procesados
Total de registros procesados: 1429


## 4. Validaciones posteriores a la insercion


In [5]:
with engine.connect() as connection:
    total_rows = connection.execute(text("SELECT COUNT(*) FROM convocatorias")).scalar_one()
    rows_with_embeddings = connection.execute(
        text("SELECT COUNT(*) FROM convocatorias WHERE embedding IS NOT NULL")
    ).scalar_one()
    sample_rows = pd.read_sql(
        text(
            """
            SELECT document_id, cuce, entidad, modalidad, estado
            FROM convocatorias
            ORDER BY id
            LIMIT 5
            """
        ),
        connection,
    )

print("Filas totales en PostgreSQL:", total_rows)
print("Filas con embedding:", rows_with_embeddings)
sample_rows


Filas totales en PostgreSQL: 1429
Filas con embedding: 1429


,document_id,cuce,entidad,modalidad,estado
0,26-1101-00-1669685-1-1,26-1101-00-1669685-1-1,Gobierno Autonomo Municipal De Sucre,ANPP,Vigente
1,26-0417-03-1669697-1-1,26-0417-03-1669697-1-1,Caja Nacional De Salud Regional Santa Cruz,CNC,Vigente
2,26-0417-03-1669695-1-1,26-0417-03-1669695-1-1,Caja Nacional De Salud Regional Santa Cruz,CNC,Vigente
3,26-0418-07-1669627-1-1,26-0418-07-1669627-1-1,Caja Petrolera De Salud Administracion Departa...,ANPP,Vigente
4,26-1205-00-1647943-2-1,26-1205-00-1647943-2-1,Gobierno Autonomo Municipal De El Alto,ANPE,Vigente


## 5. Busqueda semantica de ejemplo


In [6]:
semantic_examples = [
    "medicamentos para hospital",
    "mantenimiento de vias urbanas",
    "reactivos de laboratorio",
]

for query in semantic_examples:
    print(f"\nConsulta semantica: {query}")
    results = semantic_search(query, k=3)
    if not results:
        print("Sin resultados.")
        continue
    display(pd.DataFrame(results)[["cuce", "entidad", "objeto_contratacion", "distance"]])



Consulta semantica: medicamentos para hospital


,cuce,entidad,objeto_contratacion,distance
0,26-0424-08-1669595-1-1,Caja De Salud Cordes Regional Cochabamba,Adquisición De Productos Farmacéuticos Medicam...,0.468858
1,26-0902-21-1667666-1-1,Hospital De Clinicas,"Adquisicion De Gasa Quirurgica 100 Yardas, Alg...",0.486648
2,26-0417-06-1668691-1-1,Caja Nacional De Salud (Regional Chuquisaca),"Adquisición De Metotrexato 2,5 Mg Comprimido P...",0.510121



Consulta semantica: mantenimiento de vias urbanas
Sin resultados.

Consulta semantica: reactivos de laboratorio


,cuce,entidad,objeto_contratacion,distance
0,26-0520-00-1668033-1-1,Empresa Metalurgica Vinto - Nacionalizada,Adquisicion Repuestos Para El Mantenimiento Pr...,0.684512
1,26-0520-00-1668255-1-1,Empresa Metalurgica Vinto - Nacionalizada,Adquisicion De Electrodos De Diferentes Codifi...,0.732797


## 6. Busqueda keyword de ejemplo


In [7]:
keyword_examples = [
    "medicamentos",
    "cemento",
    "laboratorio",
]

for query in keyword_examples:
    print(f"\nConsulta keyword: {query}")
    results = keyword_search(query, k=3)
    if not results:
        print("Sin resultados.")
        continue
    display(pd.DataFrame(results)[["cuce", "entidad", "objeto_contratacion"]])



Consulta keyword: medicamentos


,cuce,entidad,objeto_contratacion
0,26-0424-08-1669595-1-1,Caja De Salud Cordes Regional Cochabamba,Adquisición De Productos Farmacéuticos Medicam...
1,26-1219-00-1669640-1-1,Gobierno Autonomo Municipal De Quime,Adquisicion De Medicamentos E Insumos Para El ...
2,26-0417-03-1669687-1-1,Caja Nacional De Salud Regional Santa Cruz,Adquisicion De Medicamentos Regional Santa Cru...



Consulta keyword: cemento


,cuce,entidad,objeto_contratacion
0,26-1519-00-1669672-1-1,Gobierno Autonomo Municipal De Tupiza,Mantenimiento Vías Urbanas Ciudad De Tupiza (A...
1,26-1206-00-1669459-1-1,Gobierno Autonomo Municipal De Viacha,"Refacción Y Mantenimiento De Caminos, Secciona..."
2,26-1206-00-1669479-1-1,Gobierno Autonomo Municipal De Viacha,Mantenimiento Alcantarillado Sanitario Y Pluvi...



Consulta keyword: laboratorio


,cuce,entidad,objeto_contratacion
0,26-0041-15-1669677-1-1,Servicio Nacional De Sanidad Agropecuaria E In...,Adquisicion De Placas Petrifilm Para El Recuen...
1,26-0041-15-1669573-1-1,Servicio Nacional De Sanidad Agropecuaria E In...,Adquisición Kits De Anticuerpos Para Detección...
2,26-0902-21-1669603-1-1,Hospital De Clinicas,Adquisicion De Reactivos Para Contador Hematol...


## 7. Observaciones

- Este notebook reutiliza la capa `src.vector_store` para evitar duplicacion de logica.
- El siguiente paso metodologico es construir el dataset de consultas de evaluacion (`dev`, `val`, `test`) y medir `Precision@K`, `Recall@K` y `MRR`.
- Si la conexion local falla, verificar Docker, `DATABASE_URL` y que el puerto `5432` este accesible desde el entorno Python usado por el notebook.
